File: subsystems/scan-engine/src/scan-progress.ts
/**
 * ==========================================================
 * SKOS
 * Smaily Knowledge Operating System
 * ==========================================================
 *
 * Subsystem : Scan Engine
 * Module    : Scan Progress
 *
 * Build     : BUILD-000169
 * Sprint    : Phase 3
 * Version   : 0.1.0
 *
 * Status    : Foundation
 * ==========================================================
 */

import {
    ScanTaskStatus
} from "./scan-task";

/**
 * Scan progress information.
 */
export interface ScanProgress {

    /**
     * Scan task identifier.
     */
    taskId: string;

    /**
     * Current task status.
     */
    status: ScanTaskStatus;

    /**
     * Total discovered files.
     */
    totalFiles: number;

    /**
     * Processed files.
     */
    processedFiles: number;

    /**
     * Current file URI or path.
     */
    currentItem?: string;

    /**
     * Last completed checkpoint.
     */
    checkpoint?: string;

    /**
     * Progress percentage.
     */
    percentage: number;

    /**
     * Last update time.
     */
    updatedAt: number;

}

/**
 * Progress Manager.
 */
export class ScanProgressManager {

    private readonly progressMap =
        new Map<string, ScanProgress>();

    /**
     * Register progress.
     */
    public register(
        progress: ScanProgress
    ): void {

        this.progressMap.set(

            progress.taskId,

            progress

        );

    }

    /**
     * Return progress.
     */
    public get(
        taskId: string
    ): ScanProgress | undefined {

        return this.progressMap.get(taskId);

    }

    /**
     * Update progress.
     */
    public update(

        taskId: string,

        processedFiles: number,

        currentItem?: string,

        checkpoint?: string

    ): boolean {

        const progress =

            this.progressMap.get(taskId);

        if (!progress) {

            return false;

        }

        progress.processedFiles =
            processedFiles;

        progress.currentItem =
            currentItem;

        progress.checkpoint =
            checkpoint;

        progress.updatedAt =
            Date.now();

        if (progress.totalFiles > 0) {

            progress.percentage =

                Math.min(

                    100,

                    (processedFiles /
                        progress.totalFiles) * 100

                );

        }

        return true;

    }

    /**
     * Mark task completed.
     */
    public complete(
        taskId: string
    ): boolean {

        const progress =

            this.progressMap.get(taskId);

        if (!progress) {

            return false;

        }

        progress.status =
            ScanTaskStatus.Completed;

        progress.percentage = 100;

        progress.updatedAt =
            Date.now();

        return true;

    }

    /**
     * Remove progress.
     */
    public remove(
        taskId: string
    ): boolean {

        return this.progressMap.delete(taskId);

    }

}

File: SKOS-Android/app/src/main/java/com/smaily/skos/scanner/ScanProgress.kt
package com.smaily.skos.scanner

/**
 * ---------------------------------------------------------
 * SKOS - Smaily Knowledge Operating System
 * Scan Progress
 * ---------------------------------------------------------
 *
 * Immutable snapshot of the current scan progress.
 *
 * Author : Jamshid Smaily Bazghaleh
 * Architecture : SKOS Professional Alpha
 * Version : 1.0.0
 * ---------------------------------------------------------
 */
data class ScanProgress(

    /**
     * Current scan stage.
     */
    val stage: ScanStage = ScanStage.IDLE,

    /**
     * Total number of files expected.
     * صفر یعنی هنوز تعداد کل مشخص نشده است.
     */
    val totalFiles: Long = 0,

    /**
     * Number of processed files.
     */
    val processedFiles: Long = 0,

    /**
     * Number of successfully processed files.
     */
    val successfulFiles: Long = 0,

    /**
     * Number of failed files.
     */
    val failedFiles: Long = 0,

    /**
     * Number of discovered directories.
     */
    val directories: Long = 0,

    /**
     * Total scanned bytes.
     */
    val processedBytes: Long = 0,

    /**
     * Mission start time.
     */
    val startedAt: Long = System.currentTimeMillis(),

    /**
     * Last update time.
     */
    val updatedAt: Long = System.currentTimeMillis()
) {

    /**
     * Progress percentage.
     */
    val percentage: Int
        get() {

            if (totalFiles <= 0L)
                return 0

            return ((processedFiles * 100L) / totalFiles)
                .coerceIn(0L, 100L)
                .toInt()
        }

    /**
     * Remaining files.
     */
    val remainingFiles: Long
        get() = (totalFiles - processedFiles).coerceAtLeast(0)

    /**
     * Elapsed time.
     */
    val elapsedMillis: Long
        get() = updatedAt - startedAt

    /**
     * Average files per second.
     */
    val filesPerSecond: Double
        get() {

            if (elapsedMillis <= 0L)
                return 0.0

            return processedFiles.toDouble() /
                    (elapsedMillis.toDouble() / 1000.0)
        }

    /**
     * Returns true if scan has finished.
     */
    fun isFinished(): Boolean {

        return stage == ScanStage.COMPLETED ||
                stage == ScanStage.CANCELLED ||
                stage == ScanStage.FAILED
    }
}
